tensorboard/webapp/metrics/views/main_view/common_selectors.ts

```diff
@@ -86,7 +106,142 @@ export const getSortedRenderableCardIdsWithMetadata = createSelector<
   });
 });
 
+const utils = {
+  getRenderableRuns(experimentIds: string[]) {
+    return createSelector(
+      getRunsFromExperimentIds(experimentIds),
+      getExperimentNames(experimentIds),
+      getCurrentRouteRunSelection,
+      getRunColorMap,
+      getExperimentIdToExperimentAliasMap,
+      (runs, experimentNames, selectionMap, colorMap, experimentIdToAlias) => {
+        return runs.map((run) => {
+          const hparamMap: RunTableItem['hparams'] = new Map();
+          (run.hparams || []).forEach((hparam) => {
+            hparamMap.set(hparam.name, hparam.value);
+          });
+          const metricMap: RunTableItem['metrics'] = new Map();
+          (run.metrics || []).forEach((metric) => {
+            metricMap.set(metric.tag, metric.value);
+          });
+          return {
+            run,
+            experimentName: experimentNames[run.experimentId] || '',
+            experimentAlias: experimentIdToAlias[run.experimentId],
+            selected: Boolean(selectionMap && selectionMap.get(run.id)),
+            runColor: colorMap[run.id],
+            hparams: hparamMap,
+            metrics: metricMap,
+          };
+        });
+      }
+    );
+  },
+
+  filterRunItemsByRegex(
+    runItems: RunTableItem[],
+    regexString: string,
+    shouldIncludeExperimentName: boolean
+  ) {
+    if (!regexString) {
+      return runItems;
+    }
+
+    return runItems.filter((item) => {
+      return matchRunToRegex(
+        {
+          runName: item.run.name,
+          experimentAlias: item.experimentAlias,
+        },
+        regexString,
+        shouldIncludeExperimentName
+      );
+    });
+  },
+
+  matchFilter(
+    filter: DiscreteFilter | IntervalFilter,
+    value: number | DiscreteHparamValue | undefined
+  ): boolean {
+    if (value === undefined) {
+      return filter.includeUndefined;
+    }
+    if (filter.type === DomainType.DISCRETE) {
+      // (upcast to work around bad TypeScript libdefs)
+      const values: Readonly<Array<typeof filter.filterValues[number]>> =
```

Comment:
this type is confusing me. What does the "number" thing do? DiscreteHparamValues are supposed to be arrays of numbers, strings, or booleans.

Reply:
This is confusing to me too (I didn't write it) but it seems to work and I don't really want to change it.